Nav_bar refactoring

pro100svitlo opened the pull request:
### What 🕵️ 🔍

- Add debounce to the nav_bar_container.dart when reset or update methods are called
- remove the `postFrameCallback`s
- `NavBarConfigMixin` return a nullable value (for cases when we might want to show NavBar but also might need to ignore it)
- minor refactor `NavBarConfig`
  - from now we keep in the memory list of mixins
  - every time when `update` method is called, we go through the list and pick up the first `config` where `type != ignored`
  - we need it, cos sometimes _screen might change their mind_ and decide that they do not have content for the `NavBar` and then previous in tree `config` should be shown (please see `nav_bar_ignored_screen.dart` for more info)
  - if no configs are found (empty mixin list or 0 mixin return `config`) - the `navBar` will be hidden
  - if the screen do not have `config` to show, but still want to avoid the previous `NavBar` to be shown - `NavBarConfig.hidden()` should be used

----------

### How to test 🥼 🔬

- run the example app
- check example app

----------

frankpepermans commented:
@pro100svitlo 🙏 use debounce from rxdart

pro100svitlo committed:
Make NavBarConfigMixin return nullable config

- update example app screens
- rename NavBarConfig constructor from `hide` to `hidden` to be more clear

pro100svitlo committed:
Rename screen states in the example app

pro100svitlo committed:
Update nav_bar_container.dart with debounce and tiny refactoring

- add debounce when call `reset` or `update` bar
- improve the way how mixin with configs are used

pro100svitlo committed:
Improve nav_bar observer.dart

- call _resetNavBarConfig with false when removing screen

pro100svitlo committed:
Fix xayn_design export for the test related code

MichaelKMalak commented on lib/src/widget/nav_bar/utils/config_mixin.dart:
```diff
@@ -1,5 +1,10 @@
+  /// If value is  not null - config will be shown.
+  /// If need to hide previous config, and show nothing
+  ///   (hide [NavBar] while widget shown) - use [NavBarConfig.hidden]
+  NavBarConfig? get navBarConfig;
```
Just a question, why make it nullable and depend on `null` when we can make a default like `NavBarConfig.default()` that when there, we debounce/get the previous in tree?

pro100svitlo replied:
That was my initial goal, to be honest 😄 
But then what is the difference between `.ignorred()` and `null` literally?

Anyway, if that makes more sense to you, sure, I will update it 😉

MichaelKMalak replied:
If it's possible to add `ignored`, it'll be much cleaner and readable

MichaelKMalak reviewed (approved):
Looks good!
left one question but not a stopper.

pro100svitlo committed:
Update nav_bar_config.dart

- add hidden and ignored type
- create ignored constructor for  NavBarConfig
- add NavBarType extension with booleans

pro100svitlo committed:
Make config_mixin.dart return non-nullable NavBarConfig

- update nav_bar.dart and nav_bar_container.dart with the new types
- update example

joecks commented on lib/src/widget/nav_bar/widget/nav_bar.dart:
```diff
@@ -41,15 +41,26 @@ class _NavBarState extends State<NavBar> implements ConfigUpdater {
   void update(NavBarConfig? config) {
+    assert(
+      config?.type != NavBarType.ignored,
+      'This config should be ignored and should NOT be populated here',
```
💅  This config is ignored, so it can not be used to update the NavBar. This is likely a internal Bug.

pro100svitlo replied:
updated

pro100svitlo commented on lib/src/widget/nav_bar/data/nav_bar_config.dart:
```diff
@@ -1,10 +1,13 @@
 enum NavBarType {
   card,
   backBtn,
+  hidden,
+  ignored,
```
If you will take a look on [this commit](https://github.com/xaynetwork/xayn_discovery_app/pull/49/commits/b58e9f3b84506c83aed467672d428207c7d19b9e) from the `xayn_discovery` app, you might understand a bit better the scenario which I was trying to handle.

Later today, on the call with the UI team,  we find out that the current implementation of the `reader mode` was wrong. It means we do not need `ignored` anymore. For now. 
But later we might need it again. So I would leave it for now.

>Didn't we discuss that the last NavBarConfig is the only relevant one?

When we have a talk, I was not aware of all the cases, as one with the reader mode.

About `FocusNode` - I do not remember that, but might be.
Anyway, just checked doc:
```
A node that returns true for hasFocus will receive key events if none of its focused descendants returned true from their onKey handler.
```

Looks like they also walk through the widget tree, look for the latest node, that returns `true`.
This is very similar to my approach: find the latest one, where `type != ignored`.

pro100svitlo committed:
Update exception msg in the nav_bar.dart
